adapter/memquota2/memquota.go

```diff
@@ -0,0 +1,246 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package memquota provides a simple in-memory quota implementation. It's
+// trivial to set up, but it has various limitations:
+//
+// - Obviously, the data set must be able to fit in memory.
+//
+// - When Mixer crashes/restarts, all quota values are erased.
+// This means this isn't good for allocation quotas although
+// it works well enough for rate limits quotas.
+//
+// - Since the data is all memory-resident and there isn't any cross-node
+// synchronization, this adapter can't be used in an Istio mixer where
+// a single service can be handled by different mixer instances.
+package memquota
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"istio.io/mixer/adapter/memquota2/config"
+	"istio.io/mixer/pkg/adapter"
+	"istio.io/mixer/pkg/status"
+	"istio.io/mixer/template/quota"
+)
+
+type builder struct {
+	types map[string]*quota.Type
+}
+
+type handler struct {
+	// common info among different quota adapters
+	common dedupUtil
+
+	// the counters we track for non-expiring quotas, protected by lock
+	cells map[string]int64
+
+	// the rolling windows we track for expiring quotas, protected by lock
+	windows map[string]*rollingWindow
+
+	// the limits we know about
+	limits map[string]config.Params_Quota
+}
+
+// ensure our types implement the requisite interfaces
+var _ quota.HandlerBuilder = &builder{}
+var _ quota.Handler = &handler{}
+
+///////////////// Configuration Methods ///////////////
+
+func (b *builder) Build(cfg adapter.Config, env adapter.Env) (adapter.Handler, error) {
+	c := cfg.(*config.Params)
+	return b.buildWithDedup(c, env, time.NewTicker(c.MinDeduplicationDuration))
+}
+
+func (b *builder) buildWithDedup(c *config.Params, env adapter.Env, ticker *time.Ticker) (*handler, error) {
+	limits := make(map[string]config.Params_Quota, len(c.Quotas))
+	for _, l := range c.Quotas {
+		limits[l.Name] = l
+	}
+
+	for k := range b.types {
+		if _, ok := limits[k]; !ok {
+			return nil, fmt.Errorf("did not find limit defined for quota %s", k)
+		}
+	}
+
+	h := &handler{
+		common: dedupUtil{
+			recentDedup: make(map[string]dedupState),
+			oldDedup:    make(map[string]dedupState),
+			ticker:      ticker,
+			getTime:     time.Now,
+			logger:      env.Logger(),
+		},
+		cells:   make(map[string]int64),
+		windows: make(map[string]*rollingWindow),
+		limits:  limits,
+	}
+
+	env.ScheduleDaemon(func() {
+		for range h.common.ticker.C {
+			h.common.Lock()
+			h.common.reapDedup()
+			h.common.Unlock()
+		}
+	})
+
+	return h, nil
+}
+
+func (b *builder) ConfigureQuotaHandler(types map[string]*quota.Type) error {
+	b.types = types
+	return nil
+}
+
+////////////////// Runtime Methods //////////////////////////
+
+func (h *handler) HandleQuota(context context.Context, instance *quota.Instance, args adapter.QuotaRequestArgs) (adapter.QuotaResult2, error) {
+	q := h.limits[instance.Name]
```

Comment:
What if the instance name isn't found at runtime?

Reply:
There's a check in the Build method to ensure we have a limit for every instance we're receiving.

In general, I don't think we need adapters to double-check that Mixer isn't doing something invalid such as sending instances the adapters aren't prepared to handle.